mpi.h

```diff
@@ -577,10 +577,18 @@ typedef void (MPI_T_event_free_cb_function)(MPI_T_event_registration event_regis
 typedef void (MPI_T_event_dropped_cb_function)(MPI_Count count, MPI_T_event_registration event_registration, int source_index, MPI_T_cb_safety cb_safety, void *user_data);
 
 /* MPI global variables */
-extern MPI_Fint* MPI_F_STATUS_IGNORE;
-extern MPI_Fint* MPI_F_STATUSES_IGNORE;
-extern MPI_F08_status* MPI_F08_STATUS_IGNORE;
-extern MPI_F08_status* MPI_F08_STATUSES_IGNORE;
+extern void*                           MPIX_F_BOTTOM;
+extern void*                           MPIX_F_IN_PLACE;
+extern void*                           MPIX_F_BUFFER_AUTOMATIC;
+extern char**                          MPIX_F_ARGV_NULL;
+extern char***                         MPIX_F_ARGVS_NULL;
+extern MPI_Fint*                       MPIX_F_ERRCODES_IGNORE;
+extern MPI_Fint*                       MPI_F_STATUS_IGNORE;
+extern MPI_Fint*                       MPI_F_STATUSES_IGNORE;
+extern MPI_F08_status*                 MPI_F08_STATUS_IGNORE;
+extern MPI_F08_status*                 MPI_F08_STATUSES_IGNORE;
+extern MPI_Fint*                       MPIX_F_UNWEIGHTED;
+extern MPI_Fint*                       MPIX_F_WEIGHTS_EMPTY;
```

Comment:
Even for `MPI_F_STATUS_IGNORE` and `MPI_F08_STATUS_IGNORE`, is there any confirmed use case? I am also curious is there any real use case for `_f2c/c2f` conversions? I only can imagine a use case similar to PETSc where a C program needs to pass a communicator object to a Fortran library.

Reply:
Looking at the example of PETSc, it is a C library with API accepting parameters such as `MPI_Comm`. So to support usage from Fortran program, it needs `MPI_Comm_f2c` to accept a communicator from Fortran side. But doing so, it is has to assume the Fortran program does not use `mpi_f08`. Is there mechanism from C side to tell whether it is being called from an Fortran code that uses `mpi.mod` or `mpi_f08.mod`? Thus, I think the current picture of even those `_f2c/c2f` APIs are broken. Rather than supporting and enlarging this mess, I think a sensible direction is to advise users not to use them. PETSc can work around by hide the direct MPI_Comm usage. It is more complex for some usages, but it will be more robust.

Reply:
I think it must be brain-dead to design an API that need accept MPI_Status directly cross language.

Reply:
In conclusion, having any Fortran inter-op constants or routines on the C side is problematic because Fortran does not have a reliable ABI and MPI does not have a standard Fortran standard (`mpi.mod` vs `mpi_f08.mod`). The interop standard on the Fortran side is fine since we can fix a C ABI.

Reply:
> But if C side tries to specify some Python or Julia interop, then that'll be a huge mess.

Definitely. That's the reason I'm here helping to fix C, rather than shoving my silly Python stuff on everyone's throat by getting Python things in the standard.

However, regarding Fortran, the damage has already been done. In fact, many API details of MPI are just bad stuff from 30 years ago that was done that way to ease Fortran interoperability. And we are still here, wasting our time, arguing about Fortran, and not being able to get things done due to its limitations.

> There are some worry from Julia in mixing Fortran library and C ABI. If C ABI does not contain _any_ Fortran stuff, then there will never be any issues.

You may be misreading the problem. What Julia folks want (and Python folks too, for sure), is to be able to swap MPI implementations at runtime for Fortran libraries. They are effectively pressing for a Fortran ABI.
The primary motivation of these new `MPI_F_*` constants came from the possibility of using them to produce a Fortran ABI (within the constraint of using the same compiler). Jeff catched the quickly and liked it. Then he spent one afternoon playing with it, and learned the hard way that its 2024 and Fortran-C interoperability still comes limitations and platform issues. 

At this point, after Jeff's linker experiments, I don't think the new `MPI_F_*` will provide any help towards a Fortran ABI. I really hope I can be proven wrong in the next few days.
IIUC, the only use case left for the constants is tools and intercepting Fortran calls from C and mess around. That's a user case that, despite maybe being valid and worth supporting, I'm not going to care about or advocate for.

Reply:
> Even for `MPI_F_STATUS_IGNORE` and `MPI_F08_STATUS_IGNORE`, is there any confirmed use case? 

Yes. Tools need it. It was discussed on the ABI PR or Slack by Bill Williams. They said support for other sentinels was useful, but less critical.

For example, if a tool wants to use status to get the bytes delivered, but use passes ignore, the tool needs to add its own status; otherwise it doesn't. If the tool intercepts the Fortran API but is written in C, it needs the address of the sentinel. 

I can see the use case for BOTTOM and INPLACE.  One might want to detect these to understand MPI usage patterns. 

I'm not sure we need one for WEIGHT sentinels but the consistency argument says we should have those.

Reply:
Assuming the tool is built in C, if Fortran bindings are built on top of C ABI, then the tools only need intercept on the C side, that is, `MPI_F_STATUS_IGNORE` should never get to the C side for the tool to worry. This is the case for MPICH. 

If the tool is built in Fortran, they should not touch the C side anyway.

If the tool have both Fortran and C component, then they can choose to interop within themselves and have the flexibility of intercept either on the Fortran side or the C side.

Thus, I am not understanding the Tools issue.

Reply:
All these pointer constants for Fortran need to be actual Fortran objects. The reason for them to exist is because Fortran doesn't support abstract pointers -- unless requiring iso_c_binding, which may never reach due to Fortran's legacy issue. 

The way MPICH support `MPI_F_STATUS_IGNORE` is via an `init` call  from Fortran to pass these objects from Fortran side to C side. So they are not constants at all from the C linker view.

Reply:
Can you write a function in C with name `mpi_send_` to intercept and mess with a Fortran call? I think you definitely can. Within this C function, how would you detect the use of the Fortran-side `MPI_BOTTOM` object? You cannot do it currently, unless MPI initialized a special pointer `MPI_F_BOTTOM` with the address for the Fortran-side `MPI_BOTTOM` object. 

I'm not saying that any would make sense or whether it is smart for tools to do this. But if someone wants to do it anyway, currently they cannot because Fortran-side things like `MPI_BOTTOM` are inaccessible to C. The adition of `MPI_F_BOTTOM` would make all these Fortran things accessible from C, of course not as a compile time constant, but as global variable, just like `MPI_F_STATUS_INGNORE`. Again, I'm not advocating for it, just stating some facts about the intention.

Reply:
> You may be misreading the problem. What Julia folks want (and Python folks too, for sure), is to be able to swap MPI implementations at runtime for Fortran libraries. They are effectively pressing for a Fortran ABI.

Fortran ABI is enforced at the compiler level. As long as distributions make sure all fortran libraries are built using consistent Fortran compilers, and the Fortran bindings are built on top of MPI ABI, I think it is swappable. Those pointer constants are actual Fortran objects. Dynamic linker should resolve them at the swap. By the way, I am referring to `MPI_STATUS_IGNORE`, not `MPI_F_STATUS_IGNORE`. `MPI_F_STATUS_IGNORE` is an internal symbol within the Fortran binding and it is not used by Fortran programs.

Reply:
It's doable as long as every implementation gets on board and is willing to accept writing their Fortran wrapper code in a specific way. For example, MPICH will have to stop using all these auxiliary internal Fortran modules that right now populate the include directory.

Reply:
> Can you write a function in C with name `mpi_send_` to intercept and mess with a Fortran call? I think you definitely can. Within this C function, how would you detect the use of the Fortran-side `MPI_BOTTOM` object? You cannot do it currently, unless MPI initialized a special pointer `MPI_F_BOTTOM` with the address for the Fortran-side `MPI_BOTTOM` object.
> 
That is how MPICH does it currently for the F77/F90 bindings. BUT it is implementation detail and it is contained inside `libmpifort.so`. Why would users care and want to restrict the implementation? The implementation is hacky and we want to hide them from users.

Reply:
Look, Fortran, especially F77, is supposed to be simple for the user. It does not support pointers, thus it does not have the messy issues C has. The complication is that we choose to implement a Fortran library on top of a C implementation, thus we are paying for the complication due to this implementation decision. All this complication is implementation details and high quality implementations should make sure not to spill it to the users. Thus, don't standardize the mess and complicate users, rather, require implementations to aim for high quality and contain their mess.

Reply:
> It's doable as long as every implementation gets on board and is willing to accept writing their Fortran wrapper code in a specific way. For example, MPICH will have to stop using all these auxiliary internal Fortran modules that right now populate the include directory.

`use mpi_f08` allows us to write the Fortran binding without the C global and Fortran common blocks, and you want us to abandon it?

Reply:
I'm going to make a very simple question: Do you favor or not the standardization of a Fortran ABI that would allow third-party MPI libraries depending on the Fortran MPI library to swap MPI implementations at runtime?
From your previous words, looks like the answer is "no, I do not support Fortran ABI". If that is the case, I have not idea why I'm here fighting with you. If a Fortran ABI has to be produced for the C ABI to be accepted, then I'll help Jeff to get it done. If a Fortran ABI is not required to get the C ABI accepted, I truly don't give a f... about the Fortran ABI.

Reply:
> If a Fortran ABI is not required to get the C ABI accepted, 

That's where I am standing. Not only IMO it is not required, it shouldn't be included.

Reply:
> I don't get what you are holding against internal Fortran modules.

If different implementations use different layouts for these different internal module names to define handles, constants, etc, then there is no way that fortran libraries can be interoperable and swapped at runtime, as a lot of mangled symbols will differ from one implementation to the other.

Reply:
I am in the understanding that modules are used during compilation not linking. Internal symbols should be contained and thus not causing linking issues. But I need check whether MPICH breaks this principle.

Reply:
Hui, please run `nm` on the MPICH Fortran libraries. You will see that the internal module names leak everywhere to symbols in the library due to mangling. That's all good within the MPICH real.

However, if MPI implementations were aiming at producing libraries that can be swapped at runtime, either all internal modules have to match in name an contents, or way easier, you just don't use internal modules. Whether that allows for both `use mpi` and `use mpi_f08` I don't know.

Reply:
I see. You mean symbols such as
```
0000000000067c60 T __mpi_f08_callbacks_MOD_mpi_comm_null_copy_fn
```
?
So we need to require Fortran bindings not to use internal modules so that the symbol will be sth like `__mpi_f08_MOD_mpi_comm_null_copy_fn`?

That's fixable.

Reply:
I'll know for sure after trying it. Created a tracking issue and link here: https://github.com/pmodels/mpich/issues/6948

Reply:
Now I looked at the issue, I think the only sensible solution is to release a stand-alone MPI Fortran binding, e.g. `libmpifort_abi.so`. The swapability of Fortran programs comes from that they are using this stand-alone implementation.

Reply:
1. Removing internal modules is the first step for a Fortran ABI to have a chance. I'm still not sure whether a Fortran ABI is doable without implementations sharing source code, at least for the outermost layer of the Fortran module and library routines.
2. There is still the issue of how to define exactly `MPI_BOTTOM` and friends. For example, using a scalar or a size-1 array could produce different manglings. At that point comes my original idea. Can somehow define the thing in C (as a global external symbol, not a constant) and make Fortran reference that C thing with `bind(c)`. Great idea, except that we ran into platform-dependent issues.

Reply:
> Now I looked at the issue, I think the only sensible solution is to release a stand-alone MPI Fortran binding, e.g. `libmpifort_abi.so` (or VAPPA)

And that's the root of why I insist the C ABI has to be rock solid, and we cannot have things left to open the interpretation of implementers :wink:.

Reply:
If you desire rock solid, leave as much complication out as you can get away. If you cover every corner usage, even a rock will behave like a TOFU :)

Reply:
Case in point, PYTHON never bothered with its ABI, and they endured the Python 2-3 breakages. So don't worry for Fortran :)

Reply:
see 6edec3394ef9a841fdad4910f1ee8241fab1dbd5

Reply:
> Even for `MPI_F_STATUS_IGNORE` and `MPI_F08_STATUS_IGNORE`, is there any confirmed use case?

Yes, please see https://github.com/mpi-forum/mpi-issues/issues/837#issuecomment-1971150993.

Reply:
These features were requested by tools developers, e.g., https://github.com/mpi-forum/mpi-issues/issues/837#issuecomment-1971139198.  I did not invent the need for them.